src/cpp/src/image_generation/models/unet2d_condition_model.cpp

```diff
@@ -52,67 +54,51 @@ UNet2DConditionModel& UNet2DConditionModel::reshape(int batch_size, int height,
     width /= m_vae_scale_factor;
 
-    std::map<std::string, ov::PartialShape> name_to_shape;
-
-    for (auto && input : m_model->inputs()) {
-        std::string input_name = input.get_any_name();
-        name_to_shape[input_name] = input.get_partial_shape();
-        if (input_name == "timestep") {
-            name_to_shape[input_name][0] = 1;
-        } else if (input_name == "sample") {
-            name_to_shape[input_name] = {batch_size, name_to_shape[input_name][1], height, width};
-        } else if (input_name == "time_ids" || input_name == "text_embeds") {
-            name_to_shape[input_name][0] = batch_size;
-        } else if (input_name == "encoder_hidden_states") {
-            name_to_shape[input_name][0] = batch_size;
-            name_to_shape[input_name][1] = tokenizer_model_max_length;
-        }
-    }
-
-    m_model->reshape(name_to_shape);
+    UNetInference::reshape(m_model, batch_size, height, width, tokenizer_model_max_length);
 
     return *this;
 }
 
 UNet2DConditionModel& UNet2DConditionModel::compile(const std::string& device, const ov::AnyMap& properties) {
     OPENVINO_ASSERT(m_model, "Model has been already compiled. Cannot re-compile already compiled model");
-    ov::Core core = utils::singleton_core();
-    ov::CompiledModel compiled_model;
+
+    if (device == "NPU") {
+        m_impl = std::make_shared<UNet2DConditionModel::UNetInferenceStaticBS1>();
+    } else {
+        m_impl = std::make_shared<UNet2DConditionModel::UNetInferenceDynamic>();
+    }
+
     std::optional<AdapterConfig> adapters;
     if (auto filtered_properties = extract_adapters_from_properties(properties, &adapters)) {
         adapters->set_tensor_name_prefix(adapters->get_tensor_name_prefix().value_or("lora_unet"));
         m_adapter_controller = AdapterController(m_model, *adapters, device);
-        compiled_model = core.compile_model(m_model, device, *filtered_properties);
+        m_impl->compile(m_model, device, *filtered_properties);
     } else {
-        compiled_model = core.compile_model(m_model, device, properties);
+        m_impl->compile(m_model, device, properties);
     }
-    m_request = compiled_model.create_infer_request();
+
     // release the original model
     m_model.reset();
 
     return *this;
 }
 
 void UNet2DConditionModel::set_hidden_states(const std::string& tensor_name, ov::Tensor encoder_hidden_states) {
-    OPENVINO_ASSERT(m_request, "UNet model must be compiled first");
-    m_request.set_tensor(tensor_name, encoder_hidden_states);
+    OPENVINO_ASSERT(m_impl, "UNet model must be compiled first");
+    m_impl->set_hidden_states(tensor_name, encoder_hidden_states);
 }
 
 void UNet2DConditionModel::set_adapters(const std::optional<AdapterConfig>& adapters) {
-    if (adapters) {
-        m_adapter_controller.apply(m_request, *adapters);
+    OPENVINO_ASSERT(m_impl, "UNet model must be compiled first");
+    if(adapters) {
+        OPENVINO_ASSERT(m_impl, "UNet model must be compiled first");
```

Comment:
```suggestion
```
as it's already asserted before `if (adapters)`

Reply:
addressed in this PR: https://github.com/openvinotoolkit/openvino.genai/pull/1172